### ISSUE TITLE
beast2vs1 test for EBSP fails

### ISSUE THREAD

**Issue by rbouckaert** [CompEvol/beast2#999]
The posterior of the testEBSP.xml fails for [TreePriorTest](https://github.com/CompEvol/beast2/blob/master/src/test/beast/beast2vs1/TreePriorTest.java). At some point it must have passed. When did this change and why? This might be a symptom of a bug introduced at some point. Finding out which commit made the test fail would be a good starting point.

**Comment by walterxie** [comment]
Hi Remco, after my investigation, it seemed the expectations for EBSP were totally wrong. I barely remember it was in development and might have been excluded from the test suits in the previous Ant call. I cannot figure it out why it is there now. But I apology for this mistaken in the past. 
So I ran the testEBSP.xml using BEAST 1.8.4, and corrected all exceptions in the EBSP test. But it seemed still a difference between posterior.
`Expected posterior delta mean: -3841.2851 - -3762.515419489472 <= delta stdErr: 2*(8.559 + 3.0511764457087116)`
Two logs are attached.

BEAST 1 [b1.tar.gz](https://github.com/CompEvol/beast2/files/7182927/b1.tar.gz)

BEAST 2 [b2.tar.gz](https://github.com/CompEvol/beast2/files/7182929/b2.tar.gz)

There are 5 more parameters not passed as attached:
[results.tar.gz](https://github.com/CompEvol/beast2/files/7182925/results.tar.gz)

**Comment by walterxie** [comment]
There were few problems to run BEAST 1:
1. I could not use BEAST 1.10.* in my Mac, which forces to use beagle 3, but throws exception when using it.
2. I have to downgrade to 1.8.4 without using beagle.
3. the testEBSP.xml in BEAST 1 seems difficult to converge. After I increased the chain length to 4 times of original length, the ESS of posterior is still 85 in the attached log. I am running a 100M-state long analysis, it hopefully will finish this night.

**Comment by walterxie** [comment]
I also fixed the names of parameters.

**Comment by walterxie** [comment]
Here is the long run. The stats are very similar to previous log

[b1-100M.tar.gz](https://github.com/CompEvol/beast2/files/7192972/b1-100M.tar.gz).

**Comment by walterxie** [comment]
I updated the two XMLs to use HCV dataset. The results and xmls are attached. Please use 20% burnin. It seems OK.

posterior: 
mean | -6867.1698 | -6897.0506
stderr of mean | 1.3096 | 2.5939

[b1.tar.gz](https://github.com/CompEvol/beast2/files/7193707/b1.tar.gz)

[b2.tar.gz](https://github.com/CompEvol/beast2/files/7193708/b2.tar.gz)

**Comment by walterxie** [comment]
I just found beast2 default not to use "Intervals Middle", which makes above two models different. I will re-run the tests tomorrow.

**Comment by walterxie** [comment]
The new logs at long runs are attached. Both suppose to use the same priors and models:

[b1-100M.tar.gz](https://github.com/CompEvol/beast2/files/7199014/b1-100M.tar.gz)

[b2-100M.tar.gz](https://github.com/CompEvol/beast2/files/7199015/b2-100M.tar.gz)

**Comment by rbouckaert** [comment]
@walterxie  it looks like the first population size, and the population mean are consistent between the runs, but subsequent population sizes are double in b1 compared to b2. Maybe the `factor=2` has something to do with that?

Has there been a time in the past where the test passed? Or did all commits since the beginning of time contain this faulty test? If there is a time in the past where this test passed, then it must be possible to find the time the test stopped passing and investigate the changes made in that commit that caused the change of behaviour.

**Comment by walterxie** [comment]
@rbouckaert I may be wrong, but according to Joseph's tutorial, the `factor` looks equivalent to `ploidy="2.0"`. I will go through the code to see whether I could confirm this or not.

I am suspecting the implementations between BEAST 2 and 1 may have difference. Not only popSizes, the indicators are very different as well. I have tried to make the priors as similar as I could in the above tests.

This test was incomplete but should be excluded from Ant testing. I am not sure when it started to be used.

**Comment by alexeid** [comment]
I would change the ploidy to 1 to check whether that fixes the difference.

**Comment by walterxie** [comment]
OK. I also found the EBSP analyser in both, but they required to log trees. I will try these 2 changes.

**Comment by walterxie** [comment]
I have re-designed the tests as Alexei's advised:
1. use HCV data set.
2. use `factor=1` and `ploidy =1`. I read both code, I can confirm they are the same parameter to scale the popSize.
3. run the analyser.

So the new logs, xmls and plots are attached below. 
- BEAST 1 : [b1.tar.gz](https://github.com/CompEvol/beast2/files/7207755/b1.tar.gz)
- BEAST 2 : [b2.tar.gz](https://github.com/CompEvol/beast2/files/7207756/b2.tar.gz)

I also combined the two plots into one, where X is "time before present", blue is beast1, red is beast2:
[b1vs2.pdf](https://github.com/CompEvol/beast2/files/7207759/b1vs2.pdf)

I think there are differences between BEAST 1 and 2. But not sure why it only affect the popSize near the present.

**Comment by walterxie** [comment]
Here is Joseph's plot https://bmcecolevol.biomedcentral.com/articles/10.1186/1471-2148-8-289/figures/7

**Comment by walterxie** [comment]
I found the difference between beast 1 plot and beast 2 was caused by the `MixedDistributionLikelihood` only existing in beast 1 xml prior section. 

```xml
<mixedDistributionLikelihood>
       <distribution0>
              <exponentialDistributionModel idref="demographic.populationMeanDist"/>
       </distribution0>
       <distribution1>
              <exponentialDistributionModel idref="demographic.populationMeanDist"/>
       </distribution1>
       <data>
              <parameter idref="demographic.popSize"/>
       </data>
       <indicators>
              <parameter idref="demographic.indicators"/>
       </indicators>
</mixedDistributionLikelihood>
```  

BEAST 2 used  
```xml
<prior id="popSizePrior.alltrees" name="distribution" x="@popSizes.alltrees">
      <Exponential id="popPriorDist.EBSP" mean="@populationMean.alltrees" name="distr"/>
</prior>
```  

I will try to make these two equivalent.

**Comment by rbouckaert** [comment]
The fragment indicates it is possible to assign a different prior to `popSize` entries based on the `indicators`, but since `distribution0` and `distribution1` refer to the same distribution, it should still produce the same prior as the one in BEAST 2 (assuming an exponential with mean=populationMean is used). If there is a difference, it must be something else.

The plots [b1vs2](https://github.com/CompEvol/beast2/files/7207759/b1vs2.pdf).pdf shows only a very small difference at the uncertainty intervals at the beginning -- the means are a very good match. Perhaps this has something to do with burn-in?

**Comment by walterxie** [comment]
Yes. You are correct about the `mixedDistributionLikelihood `. I also double-checked the Java code and XMLs for this and other priors except of `Coalescent`. They look same. So the only part to be unchecked is EBSP coalescent. 

If you look at the beast 1 log, the mean of the posterior of popSize mean is about 8.8, and then all 63 means of popSizes are between 7.4 and 9.2. But beast 2 log only has the mean of the 1st popSize around 8.8, and the rest of 62 means are around 4.  

If the 1st popSize starts from tips (present). I am suspecting there may be a bug in beast 2 EBSP to halve the popSizes after the 1st interval towards the root. But I do not understand why the plots are matched when the time is away from the present, if this assumption could be true. I will have a try.

**Comment by walterxie** [comment]
I ran the another comparison tests with new setting: stepwise, and not using mid points. BEAST 2's plot seemed wired:

[b2.pdf](https://github.com/CompEvol/beast2/files/7275379/b2.pdf)

But BEAST 1 looked fine:

[b1.pdf](https://github.com/CompEvol/beast2/files/7275381/b1.pdf)

**Comment by rbouckaert** [comment]
Interestingly, values on the left hand side of the graph are significantly different, but on the right hand side they look compatible.

**Comment by alexeid** [comment]
BEAST2 is definitely broken with these settings. The BEAST1 results are the familiar results for this data set.

**Comment by walterxie** [comment]
@alexeid I found another potential problem as well: after I used the beast 1 analyser to plot beast 2 log as you advised, the plot looked same to beast 1.

[b2-a1.pdf](https://github.com/CompEvol/beast2/files/7281780/b2-a1.pdf)

But this cannot confirm the bug is only in beast 2 `beast.app.tools.EBSPAnalyser`, because their plotting mechanisms are totally different. Not sure why the compatibility was not considered at all. So the difference is:

1. beast 1 uses the original log, and takes "popSize" and "indicators" as the input.
2. beast 2 has to log an extra log default to "EBSP.log", which looks like containing the 1st popSize and the coalescent intervals. Then the beast 2 analyser takes this special log as the input. Beast 1 does not have this, I cannot test beast 1 plot using beast 2 analyser, as I did above.

So in my assumption, the bug could be still somewhere in `CompoundPopulationFunction`, perhaps.

**Comment by walterxie** [comment]
I added two new xmls `examples/beast2vs1/beast1/testEBSP-b1.xml` `examples/beast2vs1/testEBSP-b2.xml` to replicate these problems. The previous 2 xmls are renamed to testEBSP-old.xml.
